File: kth_min_BST.py
#Time Complexity: O(N)

#Space Complexity: O(H) where H is the height of the tree

class Solution:
    def __init__(self):
        self.count = 0
        self.result = -1
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        if not root:
            return -1
        
        def inorder(root):
            if not root:
                return 
            
            inorder(root.left)
            self.count+=1
            if self.count == k:
                self.result = root.val
                return 
            inorder(root.right)
        inorder(root)
        return self.result

File: KthMinBST.java
//TC - O(n)
//SC - O(1)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count;
    int result;
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<>();
        
        while(root != null || !stack.isEmpty()){
            while(root != null){
                stack.push(root);
                root = root.left;
            }
            root = stack.pop();
            k--;
            if(k == 0)  return root.val;
            root = root.right;
        }
        return -1;
    }
}
